feat(theme): support keyframes through animation property

mnajdova opened the pull request:
# Keyframes

This PR aims to add possibility for defining and using keyframes in the theme. 

# API Proposal

We should add keyframes prop in the theme, that will define different keyframes in the following manner:

```
<Provider theme={{
    // ..some theme props
    keyframes: {
      spinner: () => ({
        from: {
          transform: 'rotate(0deg)',
        },
        to: {
          transform: 'rotate(360deg)',
        },
      }),
    }
  }}
> <Component /> </Provider>
```
Then, inside the components the users can access this keyframes inside the styles, using the theme property. Something like:

```
<Icon name="umbrella" size="big"
      styles={({theme}) => (theme.keyframes.spinner ? {animation: 
        `${theme.renderer.renderKeyframe(theme.keyframes.spinner, {})} 1s infinite`} : {})}
    />
```

Another approach that I have tried is to evaluate the keyframes inside the Provider component once and then use them like this: 
```
<Icon 
  name="umbrella" 
  size="big" 
  styles={({theme, props}) => ({
    animation: `${theme.keyframes.spinner} 1s infinite`})
  } 
/>
```

Altough this API is cleaner and easier for using, we are not able to provide any props to the keyframe, which is something that is not acceptable.

Please share your thoughts on this, so that we can conclude the final API.

# Update
After discussion, we decided to add animation property to the components, that would encapsulate all things regarding the keyframes and will provide easier usage. The definition of the keyframes in the theme will remain the same, but the usage will be provided in the following manner:
## using keyframe already defined in the theme
```
<Button
  circular
  content="C"
  animation={{
    name: 'spinner',
    duration: '2s',
    iterationCount: 'infinite',
  }}
/>
```
## using custom keyframe

```
const keyframe = ({ startColor, endColor }) => ({
  from: { color: startColor },
  to: { color: endColor },
})

<Icon
  name="umbrella"
  size="big"
  animation={{
    keyframe,
    keyframeParams: { startColor: 'red', endColor: 'yellow' },
    duration: '5s',
    delay: '2s',
    iterationCount: 'infinite',
    direction: 'alternate',
    fillMode: 'backwards',
    timingFunction: 'ease',
  }}
/>
```

The animation property is actually defining all animation-related css properties. The interface looks like this:

```
export type Animation = {
  keyframe?: any // custom keyframe function
  name?: string // theme defined keyframe
  keyframeParams?: object // params for the keyframe function
  delay?: string // animation-delay CSS property
  direction?: string // animation-direction CSS property
  duration?: string // animation-duration CSS property
  fillMode?: string // animation-fill-mode CSS property
  iterationCount?: string // animation-iteration-count CSS property
  playState?: string // animation-play-state CSS property
  timingFunction?: string // animation-timing-function CSS property
}
```

Commit:
-init implementation of keyframes

Commit:
-changed keyframes to be always functions

Commit:
-added keyframe description in the theming docs

Commit:
-added animation property to icon and button instead of requiring the user to render the keyframe themself

Commit:
-refactored animation to be part of the theme instead of just defining the keyframes
-added option for overriding the animation props

Commit:
-added Transition component

Commit:
-continuing running the paused animation

Commit:
-added description for the transition component

Commit:
-Transition component is now appling the style on the child component
-extracted logic for generating the animationStyle so that it can be reused in renderComponent, as well as the Transition component

Commit:
-renaming createAnimation to createAnimationStyles

levithomason commented on src/components/Transition/Transition.tsx:
```diff
@@ -0,0 +1,109 @@
+      timingFunction: this.props.timingFunction,
+    }
+
+    const animationStyle = createAnimationStyles(animation, theme)
```
To prevent potentially rendering keyframes on prop changes, let's remove keyframe params.  This will require all keyframes to be defined in the Provider and rendered ahead of time.

levithomason commented on src/components/Transition/Transition.tsx:
```diff
@@ -0,0 +1,109 @@
+     */
+    children: PropTypes.node,
+
+    /** Delay of the animation. */
```
Let's remove these descriptions or update to what fela has.  Ideally, these would give the user more detail about why/how they affect the transition.

levithomason commented on src/components/Icon/Icon.tsx:
```diff
@@ -44,6 +50,22 @@ class Icon extends UIComponent<Extendable<IconProps>, any> {
 
   static propTypes = {
+    /** Generic animation property for adding CSS animation to the component's root element */
+    animation: PropTypes.oneOfType([
```
Let's pull this to a common place.  All UI components will support this prop as well.

levithomason commented on docs/src/examples/components/Icon/Variations/IconAnimated.shorthand.tsx:
```diff
@@ -0,0 +1,47 @@
+    this.runningTimeout = setTimeout(() => {
+      this.setState({ playState: 'running' })
+    }, 20000)
+  }
```
Let's simplify this example to not use timeouts.  Also, move it to the Transition examples for now.  We won't show individual component transitions just yet.

levithomason commented on docs/src/examples/components/Button/Variations/index.tsx:
```diff
@@ -24,6 +24,11 @@ const Variations = () => (
+      title="Animated"
+      description="A button can have some animation applied on it."
+      examplePath="components/Button/Variations/ButtonAnimated"
+    />
```
remove

levithomason commented on src/themes/teams/index.ts:
```diff
@@ -34,11 +34,36 @@ Object.keys(fontIcons).forEach(iconName => {
 
+const animations = {
+  spinner: {
+    keyframe: () => ({
```
For now, we won't accept params to keyframes (to prevent rendering new animations on prop changes).   Instead, just accept an object here and use `callable()` on it internally.

levithomason reviewed (approved):
Merge after changes are addressed 👍

Commit:
Merge branch 'master' into feat/keyframes-support

# Conflicts:
#	src/components/Button/Button.tsx
#	src/components/Icon/Icon.tsx
#	src/lib/mergeThemes.ts

Commit:
-added tests
-removed keyframeParams
-added transition examples
-added animation prop to all UI components

Commit:
Merge branch 'master' into feat/keyframes-support

Commit:
-fix in the Transition component

Commit:
-removed comments

Commit:
-removed unused divs

Commit:
-reverting changes from Provider

Commit:
-replacing className with styles prop in the Attachment's shorthands

Commit:
-updating changelog
